src/hotspot/share/opto/memnode.cpp

```diff
@@ -143,31 +143,173 @@ extern void print_alias_types();
+    Node* out = base->fast_out(i);
+    if (!out->is_CallJava() || out->req() <= TypeFunc::Parms) {
+      continue;
+    }
+
+    Node* parm = out->in(TypeFunc::Parms);
+
+    if (parm == base || (parm->is_InlineType() && parm->as_InlineType()->get_oop() == base)) {
+      CallJavaNode* call = out->as_CallJava();
+      ciMethod* target = call->method();
+      if (!target->is_object_constructor()) {
+        continue;
+      }
+
+      // If the field is declared in a class that is a subclass of the one declaring the
+      // constructor, then the field is set inside the constructor, else the field must be set
+      // before the constructor invocation. E.g. A field Super.x will be set during the execution
+      // of Sub::<init>, while a field Sub.y must be set before Super::<init> is invoked.
+      if (target->holder()->is_subclass_of(field->holder())) {
+        return find_call_fallthrough_mem_output(call);
+      } else {
+        Node* res = call->in(TypeFunc::Memory);
+        assert(res != nullptr, "should have a memory input");
+        return res;
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+// Try to find a better memory input for a load from a strict final field
+static Node* optimize_strict_final_load_memory(PhaseGVN* phase, ciField* field, Node* adr, Node*& base_local) {
+  intptr_t offset = 0;
+  Node* base = AddPNode::Ideal_base_and_offset(adr, phase, offset);
+  if (base == nullptr) {
+    return nullptr;
+  }
+
+  Node* base_uncasted = base->uncast();
+  if (base_uncasted->is_Proj()) {
+    MultiNode* multi = base_uncasted->in(0)->as_Multi();
+    if (multi->is_Allocate()) {
+      // The result of an AllocateNode, try to find the constructor call
+      base_local = base_uncasted;
+      return optimize_strict_final_load_memory_from_local_object(field, base_uncasted->as_Proj());
+    } else if (multi->is_Call()) {
+      // The oop is returned from a call, the memory can be the fallthrough output of the call
+      return find_call_fallthrough_mem_output(multi->as_Call());
+    } else if (multi->is_Start()) {
+      // The oop is a parameter
+      if (phase->C->method()->is_object_constructor() && base_uncasted->as_Proj()->_con == TypeFunc::Parms) {
+        // The receiver of a constructor is similar to the result of an AllocateNode
+        base_local = base_uncasted;
+        return optimize_strict_final_load_memory_from_local_object(field, base_uncasted->as_Proj());
+      } else {
+        // Use the start memory otherwise
+        return multi->proj_out(TypeFunc::Memory);
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+// Whether a call can modify a strict final field of base, given that base is allocated inside the
+// current compilation unit, or is the first parameter when we are in a constructor
+static bool call_can_modify_local_object(ciField* field, CallNode* call, Node* base_local) {
+  // The fields can only be modified in this method or in a constructor
```

Comment:
Comments and naming of this method are a bit confusing. Isn't this basically checking if `base_local` is the receiver of a constructor call `call`?

Reply:
Yes the comment should refer to `base_local` not `base`. I have also added a section clarifying that this method is equivalent to asking whether `base_local` is the receiver of a constructor call `call` and the holder of `call` is a subclass of the holder of `field`.

Reply:
Looks good, thanks for updating.